src/django_components/dependencies.py

```diff
@@ -0,0 +1,764 @@
+# collected and automatically deleted from the dict.
+if sys.version_info < (3, 9):
+    comp_hash_mapping: WeakValueDictionary = WeakValueDictionary()
+else:
+    comp_hash_mapping: WeakValueDictionary[str, Type["Component"]] = WeakValueDictionary()
+
+
+# Convert Component class to something like `TableComp_a91d03`
+@lru_cache(None)
+def _hash_comp_cls(comp_cls: Type["Component"]) -> str:
+    full_name = get_import_path(comp_cls)
+    comp_cls_hash = md5(full_name.encode()).hexdigest()[0:6]
+    return comp_cls.__name__ + "_" + comp_cls_hash
+
+
+def _gen_cache_key(
+    comp_cls_hash: str,
+    script_type: ScriptType,
+) -> str:
+    return f"__components:{comp_cls_hash}:{script_type}"
+
+
+def _is_script_in_cache(
+    comp_cls: Type["Component"],
+    script_type: ScriptType,
+) -> bool:
+    comp_cls_hash = _hash_comp_cls(comp_cls)
+    cache_key = _gen_cache_key(comp_cls_hash, script_type)
+    return comp_media_cache.has(cache_key)
+
+
+def _cache_script(
+    comp_cls: Type["Component"],
+    script: str,
+    script_type: ScriptType,
+) -> None:
+    """
+    Given a component and it's inlined JS or CSS, store the JS/CSS in a cache,
+    so it can be retrieved via URL endpoint.
+    """
+    comp_cls_hash = _hash_comp_cls(comp_cls)
+
+    # E.g. `__components:MyButton:js:df7c6d10`
+    if script_type in ("js", "css"):
+        cache_key = _gen_cache_key(comp_cls_hash, script_type)
+    else:
+        raise ValueError(f"Unexpected script_type '{script_type}'")
+
+    # NOTE: By setting the script in the cache, we will be able to retrieve it
+    # via the endpoint, e.g. when we make a request to `/components/cache/MyComp_ab0c2d.js`.
+    comp_media_cache.set(cache_key, script.strip())
+
+
+def cache_inlined_js(comp_cls: Type["Component"], content: str) -> None:
+    if not _is_nonempty_str(comp_cls.js):
+        return
+
+    # Prepare the script that's common to all instances of the same component
+    # E.g. `my_table.js`
+    if not _is_script_in_cache(comp_cls, "js"):
+        _cache_script(
+            comp_cls=comp_cls,
+            script=content,
+            script_type="js",
+        )
+
+
+def cache_inlined_css(comp_cls: Type["Component"], content: str) -> None:
+    if not _is_nonempty_str(comp_cls.js):
+        return
+
+    # Prepare the script that's common to all instances of the same component
+    if not _is_script_in_cache(comp_cls, "css"):
+        # E.g. `my_table.css`
+        _cache_script(
+            comp_cls=comp_cls,
+            script=content,
+            script_type="css",
+        )
+
+
+#########################################################
+# 2. Modify the HTML to use the same IDs defined in previous
+#    step for the inlined CSS and JS scripts, so the scripts
+#    can be applied to the correct HTML elements. And embed
+#    component + JS/CSS relationships as HTML comments.
+#########################################################
+
+
+class Dependencies(NamedTuple):
```

Comment:
Initially I had a list of `Dependencies` objects, hence this class. However, then I figured out we don't need topass around that list, but forgot to remove the `Dependencies` class.

This will be removed in one of the subsequent MRs.